### DISCUSSION TITLE
How do I get my parser to work by matching this very particular MarkDown?

### DISCUSSION THREAD

**Discussion by inuitviking** [thephpleague/commonmark#1028] [Q&A] [answered]
Hi!

I need some help with this one, and don't really know how to continue. I would greatly appreciate any help. :smile: 

I have started creating an extension for some really specific Markdown that I have used at work. It isn't non-disclosable, as this is my personal extension that I plan on making fully public and open source plugin for my own markdown "static" site thing.

Here is the gist of what I have plotted together for the time being:
https://gist.github.com/inuitviking/bf3f0836e25a7043ef55f08f0be3a408

I have created a regex that matches with markdown that looks like this:

```markdown
[![latest release](https://gitlab.com/path/to/latest/release/badge.svg)](https://gitlab.com/path/to/download/link)
```

As per regex101.com it seems like it would match that just fine: https://regex101.com/r/K93jNP/1

But once that regex goes through the InlineParserEngine, it's as if it gets lost in there somewhere. I don't really know why, but my conclusion is, since it doesn't run the `parse()` method, the InlineParserEngine must have done one of these two:

- The regex somehow didn't match anyway
- The specific markdown has already been parsed by probably a built in parser for links and images, and thus skips further parsers

Don't mind the further methods in the DownloadLastTagParser.php in the gist, my goal is currently only to make it run the `parse()` method, then I can adjust the other methods. :sweat_smile:

Again, I would be very grateful for any help on this matter. :relaxed:

**Comment by colinodell** [comment, marked as answer]
Hi there! 👋 

It looks like your Markdown syntax is already recognized and easily parsed by this library - it's just an image inside of a link.  Instead of trying to write your own parser (that needs to compete with the built-in ones), it would be *much* easier to find and modify the parsed elements.

See [these docs on how to iterate and modify the parsed elements](https://commonmark.thephpleague.com/2.4/customization/abstract-syntax-tree/) and [this example of how to hook that logic up to the engine](https://commonmark.thephpleague.com/2.4/customization/event-dispatcher/#example).

In your case, it looks like you'll want to find any `Link` containing a single `Image` child whose `Text` is one of those four strings.  You can then choose to modify or completely replace any/all of those elements as needed.

Hope that helps!

**Comment by inuitviking** [comment]
Thanks @colinodell !

I haven't tried it yet, but reading through it, it seems to make sense, and should be somewhat easy to implement. :smile: 

**Update**: I completely forgot about this one, and found it in my old notifications. What was suggested worked like a dream. Listening for the DocumentParsed event, and finding the particular part I needed was much easier, and somewhat straight forward to implement. It worked, and there is no competing with parsers. :smile:
